java/ql/src/experimental/Security/CWE/CWE-094/SpringViewManipulationLib.qll

```diff
@@ -0,0 +1,104 @@
+/**
+ * Provides classes for reasoning about Spring View Manipulation vulnerabilities
+ */
+
+import java
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.dataflow.TaintTracking
+import semmle.code.java.frameworks.spring.Spring
+import SpringFrameworkLib
+
+predicate thymeleafIsUsed() {
+  exists(Pom p, Artifact a | a = p.getADependency().getArtifact() |
+    a.getValue() =
+      [
+        "spring-boot-starter-thymeleaf", "thymeleaf-spring4", "springmvc-xml-thymeleaf",
+        "thymeleaf-spring5"
+      ]
+  )
+  or
+  exists(SpringBean b | b.getClassNameRaw().matches("org.thymeleaf.spring%"))
+}
+
+/**
+ * A taint-tracking configuration for unsafe user input
+ * that can lead to Spring View Manipulation vulnerabilities.
+ */
+class SpringViewManipulationConfig extends TaintTracking::Configuration {
+  SpringViewManipulationConfig() { this = "Spring View Manipulation Config" }
+
+  override predicate isSource(DataFlow::Node source) {
+    source instanceof RemoteFlowSource or
+    source instanceof WebRequestSource
+  }
+
+  override predicate isSink(DataFlow::Node sink) { sink instanceof SpringViewManipulationSink }
+
+  override predicate isSanitizer(DataFlow::Node node) {
+    // Block flows like
+    // ```
+    // a = "redirect:" + taint`
+    // ```
+    exists(AddExpr e, StringLiteral sl |
+      node.asExpr() = e.getControlFlowNode().getASuccessor*() and
+      sl = e.getLeftOperand*() and
+      sl.getRepresentedString().matches(["redirect:%", "ajaxredirect:%", "forward:%"])
+    )
+    or
+    // Block flows like
+    // ```
+    // x.append("redirect:");
+    // x.append(tainted());
+    // return x.toString();
+    //
+    // "redirect:".concat(taint)
+    //
+    // String.format("redirect:%s",taint);
+    // ```
+    exists(Interface appendableInterface, Method c, Call ca, StringLiteral sl |
+      (
+        exists(RefType t |
+          appendableInterface.hasQualifiedName("java.lang", "Appendable") and
+          t.extendsOrImplements*(appendableInterface) and
+          c.getDeclaringType() = t and
+          c.hasName("append")
+        )
+        or
+        c.getDeclaringType().hasQualifiedName("java.lang", "String") and
+        c.hasName(["concat", "format"])
```

Comment:
`appendableInterface` is unconstrainted in this disjunct! Shrink its scope to just the other disjunct